src/sinks/redis.rs

```diff
@@ -0,0 +1,485 @@
+                    error!(message = "Redis sink generated an error.", %error);
+                    emit!(RedisSendEventFailed { error });
+                    return Poll::Ready(Err(()));
+                }
+                None => break,
+            }
+        }
+        Poll::Ready(Ok(()))
+    }
+
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.poll_flush(cx)
+    }
+}
+
+async fn healthcheck(mut conn: ConnectionManager) -> crate::Result<()> {
+    redis::cmd("PING")
+        .query_async(&mut conn)
+        .await
+        .map_err(Into::into)
+}
+
+fn encode_event(mut event: Event, encoding: &EncodingConfig<Encoding>) -> String {
+    encoding.apply_rules(&mut event);
+
+    match encoding.codec() {
+        Encoding::Json => serde_json::to_string(event.as_log())
+            .map_err(|error| emit!(RedisEncodeEventFailed { error }))
+            .unwrap_or_default(),
+        Encoding::Text => event
+            .as_log()
+            .get(log_schema().message_key())
+            .map(|v| v.to_string_lossy())
+            .unwrap_or_default(),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::collections::HashMap;
+
+    #[test]
+    fn generate_config() {
+        crate::test_util::test_generate_config::<RedisSinkConfig>();
+    }
+
+    #[test]
+    fn redis_event_json() {
+        let msg = "hello_world".to_owned();
+        let mut evt = Event::from(msg.clone());
+        evt.as_mut_log().insert("key", "value");
+        let result = encode_event(evt, &EncodingConfig::from(Encoding::Json));
+        let map: HashMap<String, String> = serde_json::from_str(result.as_str()).unwrap();
+        assert_eq!(msg, map[&log_schema().message_key().to_string()]);
+    }
+
+    #[test]
+    fn redis_event_text() {
+        let msg = "hello_world".to_owned();
+        let evt = Event::from(msg.clone());
+        let event = encode_event(evt, &EncodingConfig::from(Encoding::Text));
+        assert_eq!(event, msg);
+    }
+
+    #[test]
+    fn redis_encode_event() {
+        let msg = "hello_world";
+        let mut evt = Event::from(msg);
+        evt.as_mut_log().insert("key", "value");
+
+        let event = encode_event(
+            evt,
+            &EncodingConfigWithDefault {
+                codec: Encoding::Json,
+                except_fields: Some(vec!["key".into()]),
+                ..Default::default()
+            }
+            .into(),
+        );
+
+        let map: HashMap<String, String> = serde_json::from_str(event.as_str()).unwrap();
+        assert!(!map.contains_key("key"));
+    }
+}
+
+#[cfg(feature = "redis-integration-tests")]
+#[cfg(test)]
+mod integration_tests {
+    use super::*;
+    use crate::test_util::{random_lines_with_stream, random_string, trace_init};
+    use futures::StreamExt;
+    use rand::Rng;
+
+    const REDIS_SERVER: &str = "redis://127.0.0.1:6379/0";
+
+    #[tokio::test]
+    async fn redis_sink_list_lpush() {
```

Comment:
For these `lpush` and `rpush` tests, could we test that the events are actually pushed in the right order? Or is that happening already and I just don't see it?

Reply:
add this code to test events are actually pushed in the right order
```rust
……
let num_events = rng.gen_range(10000..20000);
let mut events: Vec<Event> = Vec::new();
for i in 0..num_events {
    let s: String = i.to_string();
    let e = Event::from(s);
    events.push(e);
}

let stream = stream::iter(events.clone());
stream.map(Ok).forward(sink).await.unwrap();
……
for i in 0..num_events {
    let e = events.get(i).unwrap().as_log();
    let s = serde_json::to_string(e).unwrap_or_default();
    debug!("Raw event:{}", s);
    let payload: (String, String) = conn.blpop(key.clone(), 2000).await.unwrap();
    let val = payload.1;
    debug!("Read val:{}", val);
    assert_eq!(val, s);
}
```

Reply:
> For these `lpush` and `rpush` tests, could we test that the events are actually pushed in the right order? Or is that happening already and I just don't see it?


I have add this code to test events are actually pushed in the right order
```rust
……
let num_events = rng.gen_range(10000..20000);
let mut events: Vec<Event> = Vec::new();
for i in 0..num_events {
    let s: String = i.to_string();
    let e = Event::from(s);
    events.push(e);
}

let stream = stream::iter(events.clone());
stream.map(Ok).forward(sink).await.unwrap();
……
for i in 0..num_events {
    let e = events.get(i).unwrap().as_log();
    let s = serde_json::to_string(e).unwrap_or_default();
    debug!("Raw event:{}", s);
    let payload: (String, String) = conn.blpop(key.clone(), 2000).await.unwrap();
    let val = payload.1;
    debug!("Read val:{}", val);
    assert_eq!(val, s);
}
```